### ISSUE TITLE
Gatus leaves dangling open file descriptors after reloading configuration

### ISSUE THREAD

**Issue by heitorPB** [TwiN/gatus#536]
### Describe the bug

We noticed a weird behavior in Gatus that led to a horde of false negatives, all with the failure message related to too many open files. Curiously this happened to only a few endpoints.

After poking around, I found out that Gatus leaves open file descriptors and create new ones when re-reading the configuration files.

### What do you expect to see?

Gatus should close file descriptors instead of leaving them open after a reload.

### List the steps that must be taken to reproduce this issue

1. Add an sqlite database to the default configuration file, e.g.:
   ```yaml
   storage:
     type: sqlite
     path: /tmp/data.db
   ```
1. Start Gatus directly, i.e. `./gatus`
1. Open another terminal and list all open files from Gatus, e.g. with `watch`: 
   ```bash
   $ watch -d -n .5 --color ls -lah --color /proc/`ps aux | grep './gatus' | head -n 1 | awk '{print $2}'`/fd
   ```

At this point you should have Gatus running on one terminal and another one showing all open file descriptors from Gatus. With the default configuration file + an sqlite DB I see 10 open fd:

```
lrwx------ 1 h users 64 jul 25 15:43 0 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 1 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 10 -> socket:[58680]
lrwx------ 1 h users 64 jul 25 15:43 2 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 3 -> /tmp/data.db
lrwx------ 1 h users 64 jul 25 15:43 4 -> anon_inode:[eventpoll]
lr-x------ 1 h users 64 jul 25 15:43 5 -> pipe:[53621]
l-wx------ 1 h users 64 jul 25 15:43 6 -> pipe:[53621]
lrwx------ 1 h users 64 jul 25 15:43 7 -> /tmp/data.db-wal
lrwx------ 1 h users 64 jul 25 15:43 8 -> /tmp/data.db-shm
lrwx------ 1 h users 64 jul 25 15:43 9 -> socket:[53627]
```

Now trigger a reload of the config files by renaming any endpoint in the `config.yaml`, wait for gatus to reload and check the number of open files. We should have the same number of fds, but there are more:

```
dr-x------ 2 h users  0 jul 25 15:43 .
dr-xr-xr-x 9 h users  0 jul 25 15:42 ..
lrwx------ 1 h users 64 jul 25 15:43 0 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 1 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 10 -> socket:[58680]
lrwx------ 1 h users 64 jul 25 15:47 11 -> /tmp/data.db-wal
lrwx------ 1 h users 64 jul 25 15:49 12 -> socket:[80202]
lrwx------ 1 h users 64 jul 25 15:50 13 -> socket:[80206]
lrwx------ 1 h users 64 jul 25 15:43 2 -> /dev/pts/3
lrwx------ 1 h users 64 jul 25 15:43 3 -> /tmp/data.db
lrwx------ 1 h users 64 jul 25 15:43 4 -> anon_inode:[eventpoll]
lr-x------ 1 h users 64 jul 25 15:43 5 -> pipe:[53621]
l-wx------ 1 h users 64 jul 25 15:43 6 -> pipe:[53621]
lrwx------ 1 h users 64 jul 25 15:43 7 -> /tmp/data.db-wal
lrwx------ 1 h users 64 jul 25 15:43 8 -> /tmp/data.db-shm
lrwx------ 1 h users 64 jul 25 15:43 9 -> /tmp/data.db
```

Renaming more endpoints increase the number of open fds even more.

Summary:
- every time the configuration is reloaded, gatus reopen the DB file, leaving the other file descriptors open. In the example above you can see `/tmp/data.db` listed twice. Reload the configuration again and you will see it 3 times.
- Something similar happens to the sockets, they are not closed before a reload.

### Version

Latest commit from `master` branch: 162dd9bc

### Additional information

Gatus compiled with:

```bash
$ go version
go version go1.20.4 linux/amd64

$ CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o gatus .
```

**Comment by heitorPB** [comment]
@TwiN were you able to reproduce this behavior?

**Comment by TwiN** [comment]
@heitorPB Excellent catch.
I imagine this is because we're calling `initializeStorage` in `listenToConfigurationFileChanges` when a change is detected without closing the Store.

https://github.com/TwiN/gatus/blob/87d97226215fae13872da1de0012f2713bab0fb4/main.go#L107C4-L107C21

We'd have to call `storage.Get().Close()` before re-initializing the store, I suppose?

**Comment by heitorPB** [comment]
Oh, that might be it! Hopefully I'll have some time to test it next week.

**Comment by vax-r** [comment]
@TwiN 
Hello I want to try to solve this issue
Can you assign it to me? or is it taken already?

**Comment by heitorPB** [comment]
> I can reproduce the bug, seems that we don't even need to add a new endpoint or storage Simply changing the name of the default config in the repo will make the open fd to increase
> 
> The opening fd increase by 1 each time I made some changes to the config.yaml file (I change the endpoint name)

Glad that you can reproduce it! This issue is not only not closing the `storage`, it also appears in some endpoint types. I haven't figured out yet which ones though.

> Can you assign it to me? or is it taken already?

I started poking at it, but I haven't had enough time. @vax-r, you can take this if you want, and I'll gladly test your patch when you land a PR!

**Comment by vax-r** [comment]
@heitorPB 
I saw that socket fd doesn't close either, I'm not sure whether there are more, but I'll try to poke it more these days
Let me know if you've solve it , I would like to learn how to solve it

**Comment by vax-r** [comment]
@heitorPB @TwiN 
After my experiments , I've found out that the following call to make HTTP request cause the problem of dangling socket fd
https://github.com/TwiN/gatus/blob/2be6a1e5f38dceaf2644b2542aaa7a63f98d6399/core/endpoint.go#L347

The function call will make a 'keep-alive' connection to the endpoint, that's why we can see the socket fd under the `./gatus` process

According to https://pkg.go.dev/net/http#Client.CloseIdleConnections
We can close the idle connection from this function call.

I've make sure that this function can close the dangling socket fd for us, so I'm working on it to add the function properly into the code now.
I'll open a PR once I finished it.
Hope you guys can review the PR for me

Thank you!
